transports/webrtc/src/tokio/upgrade.rs

```diff
@@ -0,0 +1,237 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod noise;
+
+use futures::channel::oneshot;
+use futures::future::Either;
+use futures_timer::Delay;
+use libp2p_core::{identity, PeerId};
+use rand::distributions::Alphanumeric;
+use rand::{thread_rng, Rng};
+use webrtc::api::setting_engine::SettingEngine;
+use webrtc::api::APIBuilder;
+use webrtc::data::data_channel::DataChannel;
+use webrtc::data_channel::data_channel_init::RTCDataChannelInit;
+use webrtc::dtls_transport::dtls_role::DTLSRole;
+use webrtc::ice::network_type::NetworkType;
+use webrtc::ice::udp_mux::UDPMux;
+use webrtc::ice::udp_network::UDPNetwork;
+use webrtc::peer_connection::configuration::RTCConfiguration;
+use webrtc::peer_connection::RTCPeerConnection;
+
+use std::{net::SocketAddr, sync::Arc, time::Duration};
+
+use crate::tokio::{error::Error, fingerprint::Fingerprint, sdp, substream::Substream, Connection};
+
+/// Creates a new outbound WebRTC connection.
+pub async fn outbound(
+    addr: SocketAddr,
+    config: RTCConfiguration,
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+    client_fingerprint: Fingerprint,
+    server_fingerprint: Fingerprint,
+    id_keys: identity::Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    log::debug!("new outbound connection to {addr})");
+
+    let peer_connection = new_outbound_connection(addr, config, udp_mux).await?;
+
+    let offer = peer_connection.create_offer(None).await?;
+    log::debug!("created SDP offer for outbound connection: {:?}", offer.sdp);
+    peer_connection.set_local_description(offer).await?;
+
+    let answer = sdp::answer(addr, &server_fingerprint);
+    log::debug!(
+        "calculated SDP answer for outbound connection: {:?}",
+        answer
+    );
+    peer_connection.set_remote_description(answer).await?; // This will start the gathering of ICE candidates.
+
+    let data_channel = create_substream_for_noise_handshake(&peer_connection).await?;
+    let peer_id = noise::outbound(
+        id_keys,
+        data_channel,
+        server_fingerprint,
+        client_fingerprint,
+    )
+    .await?;
+
+    Ok((peer_id, Connection::new(peer_connection).await))
+}
+
+/// Creates a new inbound WebRTC connection.
+pub async fn inbound(
+    addr: SocketAddr,
+    config: RTCConfiguration,
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+    server_fingerprint: Fingerprint,
+    remote_ufrag: String,
+    id_keys: identity::Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    log::debug!("new inbound connection from {addr} (ufrag: {remote_ufrag})");
+
+    let peer_connection =
+        new_inbound_connection(addr, config, udp_mux, &server_fingerprint.to_ufrag()).await?;
```

Comment:
Note that the specification reuses the ufrag from the client.

Reply:
https://github.com/melekes/rust-libp2p/blob/532b41025924b7e6551e21ceb566df0073c8d0e3/transports/webrtc/src/tokio/upgrade.rs#L94

we also use the ufrag from the client. the line you pointed sets ufrag of the server (via `SettingsEngine`), which will be used when generating answer https://github.com/melekes/rust-libp2p/blob/532b41025924b7e6551e21ceb566df0073c8d0e3/transports/webrtc/src/tokio/upgrade.rs#L98

Reply:
> that we use the same ufrag for both the offer and the answer.

we can't use the same ufrag for both the offer and the answer. It's simply not correct according to ICE spec and won't work.

In case of incoming connection, we use client's ufrag to generate offer `sdp::offer(addr, &remote_ufrag)` and server's ufrag to generate answer `peer_connection.create_answer(None)` (note it's set via settings engine in `new_inbound_connection`).

In case of outgoing connection, we use random ufrag to generate offer `peer_connection.create_offer(None)` (note it's set via settings engine in `new_outbound_connection`) and server's ufrag to generate answer `sdp::answer(addr, &server_fingerprint);`.

Reply:
@melekes We can use the same ufrag to generate the answer. https://github.com/libp2p/go-libp2p/pull/1655/files#diff-097ca3a9a3a76a843b598d8b9b5f4ca582468b84894c7dfd2a976fd7135e88d2R172 uses the remote ufrag to generate the answer on the server.

Reply:
> We can use the same ufrag to generate the answer.

I'm confused :) How does it work? The purpose of SDP is to exchange information between this node and remote, not to use the same credentials in offer (this node) and answer (coming from remote).

Reply:
Consider the critical information required in an SDP exchange for connection establishment:
1. Remote IP + Port
2. TLS fingerprint (only for outgoing on browser)
3. Remote Ufrag
4. Remote passwd

Since we do not have an actual SDP exchange, we need to generate local sdp and infer the remote's sdp.

The client would infer the server's answer in order to send a stun packet using the following scheme:
1. Generate a random string `ufrag(c)` and set it as the local ice-pwd and ufrag.
2. Infer server IP + port from multiaddr
2. Infer server TLS fingerprint from multiaddr
3. Set `ufrag(c)` as the server ice-pwd and ufrag.

For incoming connections, the server infers the client's offer as follows:
1. IP + port, from the udp packet
2. TLS fingerprint isn't required since we disable verification
3. ufrag, passwd `ufrag(c)` from the incoming stun packet's username field.

At this point, the server can use the webrtc-rs or pion API to create a new peer connection on the muxed socket with ice credentials equal to `ufrag(c)`. This should connect on the 2nd-ish stun retry from the client, or if the packet is buffered in the connection, it should send the stun response immediately.

Reply:
Oh, so you're using the same ufrag/passwd for both client and server. It's okay, but the benefit of this implementation is persistent server's identity between restarts. Since the certificate will be persisted to disk most of the times, server's ufrag, which is derived from it's certificate, stays the same between restarts. I think it's better from debugging perspective.

Reply:
The server's identity will remain stable either way as long as the same interface and port is bound to and the certificate is fetched from disk.

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/2622/commits/b4472a85e513bd6bc5ab65e4868843573b7a83df